File: typescript-nodejs/architecture-shell-cqrs/src/behaviors/TelemetryBehavior.ts
import { BaseRequest } from '../BaseRequest';
import { Command } from '../Command';
import {
  IPipelineBehavior,
  RequestHandlerDelegate,
} from '../IPipelineBehavior';
import { IUnitOfWork } from '../IUnitOfWork';
import { ILogger, IResult } from './UnitOfWorkBehavior';

/**
 * TelemetryBehavior - Pipeline behavior that logs request execution metrics
 *
 * Captures request type, duration, status, and exceptions for observability.
 * Logs TransactionId for all command executions per NFR-002 and BR-002.
 *
 * @typeParam TRequest - The request type
 * @typeParam TResponse - The response type
 *
 * @remarks
 * Recommended order: 40 (after transaction management, wraps handler execution)
 *
 * Required telemetry fields per data-model.md:L459-L466:
 * - Timestamp: ISO 8601 datetime
 * - LogLevel: Info/Warning/Error
 * - RequestType: Fully qualified request type name
 * - Duration: Elapsed time in milliseconds
 * - Status: Success|BusinessFailure|InfrastructureError|Cancelled
 * - Exception: Full exception details if thrown
 * - TransactionId: GUID for ALL command executions (REQUIRED per NFR-002)
 *
 * @example
 * ```typescript
 * const behavior = new TelemetryBehavior(logger, unitOfWork);
 * mediator.addBehavior(behavior, new AllRequestsMatcher());
 * ```
 */
export class TelemetryBehavior<
  TRequest extends BaseRequest<TResponse>,
  TResponse
> implements IPipelineBehavior<TRequest, TResponse>
{
  readonly order = 40;

  constructor(
    private readonly logger: ILogger,
    private readonly unitOfWork?: IUnitOfWork
  ) {}

  async handle(
    request: TRequest,
    next: RequestHandlerDelegate<TResponse>,
    signal: AbortSignal
  ): Promise<TResponse> {
    const startTime = Date.now();
    const requestType = request.constructor.name;
    const timestamp = new Date().toISOString();

    // Log request start
    this.logger.info(
      {
        timestamp,
        requestType,
        transactionId: this.getTransactionId(request),
      },
      `${requestType} started`
    );

    try {
      // Execute handler
      const response = await next();
      const duration = Date.now() - startTime;

      // Detect business failure vs success
      const status = this.isBusinessFailure(response)
        ? 'BusinessFailure'
        : 'Success';

      // Log completion
      this.logger.info(
        {
          timestamp: new Date().toISOString(),
          requestType,
          duration,
          status,
          transactionId: this.getTransactionId(request),
        },
        `${requestType} completed in ${duration}ms - ${status}`
      );

      return response;
    } catch (error) {
      const duration = Date.now() - startTime;
      const status = signal.aborted ? 'Cancelled' : 'InfrastructureError';

      // Log error with full exception details
      this.logger.error(
        {
          timestamp: new Date().toISOString(),
          requestType,
          duration,
          status,
          exception: {
            type: error instanceof Error ? error.constructor.name : 'Unknown',
            message: error instanceof Error ? error.message : String(error),
            stack: error instanceof Error ? error.stack : undefined,
          },
          transactionId: this.getTransactionId(request),
        },
        `${requestType} failed after ${duration}ms`
      );

      // Re-throw to propagate exception
      throw error;
    }
  }

  /**
   * Gets TransactionId for command executions (REQUIRED per NFR-002 and BR-002)
   * Returns undefined for queries (no transaction)
   */
  private getTransactionId(request: TRequest): string | undefined {
    if (!this.isCommand(request)) {
      return undefined;
    }

    return this.unitOfWork?.hasActiveTransaction
      ? this.unitOfWork.transactionId
      : undefined;
  }

  /**
   * Detects business failures (Result.Failure) vs successes
   */
  private isBusinessFailure(response: TResponse): boolean {
    if (response === null || response === undefined) {
      return false;
    }

    // Check if response implements IResult with isFailure property
    return (
      typeof response === 'object' &&
      'isFailure' in response &&
      (response as IResult).isFailure === true
    );
  }

  /**
   * Type guard to detect command requests
   */
  private isCommand(request: any): request is Command {
    return (
      '_isCommand' in request ||
      request.constructor.name.endsWith('Command') ||
      Object.getPrototypeOf(request).constructor.name.endsWith('Command')
    );
  }
}

File: python-django/architecture-shell-cqrs/src/architecture_shell_cqrs/concrete_behaviors/telemetry_behavior.py
"""TelemetryBehavior - Request execution metrics logging."""

import time
import traceback
from datetime import datetime
from typing import TypeVar, Optional, Protocol
from architecture_core.functional import Result
from architecture_shell_cqrs.behaviors import IPipelineBehavior
from architecture_shell_cqrs.requests import ICommand
from architecture_shell_cqrs.unit_of_work import IUnitOfWork

TRequest = TypeVar("TRequest")
TResponse = TypeVar("TResponse")


class ILogger(Protocol):
    """Minimal logger interface compatible with Python logging module."""

    def error(self, message: str, **context) -> None:
        """Log error message with context."""
        ...

    def warn(self, message: str, **context) -> None:
        """Log warning message with context."""
        ...

    def info(self, message: str, **context) -> None:
        """Log info message with context."""
        ...

    def debug(self, message: str, **context) -> None:
        """Log debug message with context."""
        ...


class TelemetryBehavior(IPipelineBehavior[TRequest, TResponse]):
    """
    Pipeline behavior that logs request execution metrics.

    Captures request type, duration, status, and exceptions for observability.
    Logs transaction_id for all command executions per NFR-002 and BR-002.

    **Recommended order:** 40 (after transaction management, wraps handler execution)

    **Required telemetry fields per data-model.md:L459-L466:**
    - Timestamp: ISO 8601 datetime
    - LogLevel: Info/Warning/Error
    - RequestType: Fully qualified request type name
    - Duration: Elapsed time in milliseconds
    - Status: Success|BusinessFailure|InfrastructureError|Cancelled
    - Exception: Full exception details if thrown
    - TransactionId: UUID for ALL command executions (REQUIRED per NFR-002)

    Example:
        ```python
        behavior = TelemetryBehavior(logger, unit_of_work)
        mediator.register_behavior(behavior)
        ```
    """

    order = 40

    def __init__(self, logger: ILogger, unit_of_work: Optional[IUnitOfWork] = None):
        """
        Initialize TelemetryBehavior.

        Args:
            logger: Logger implementation for metrics
            unit_of_work: Optional UnitOfWork for transaction_id correlation
        """
        self.logger = logger
        self.unit_of_work = unit_of_work

    async def handle(self, request: TRequest, next_handler) -> Result[TResponse]:
        """
        Handle request with telemetry logging.

        Args:
            request: The request being processed
            next_handler: Delegate to invoke next behavior or handler

        Returns:
            Result containing response or error
        """
        start_time = time.time()
        request_type = type(request).__name__
        timestamp = datetime.utcnow().isoformat()

        # Log request start
        self.logger.info(
            f"{request_type} started",
            timestamp=timestamp,
            request_type=request_type,
            transaction_id=self._get_transaction_id(request),
        )

        try:
            # Execute handler
            response = await next_handler()
            duration_ms = (time.time() - start_time) * 1000

            # Detect business failure vs success
            status = (
                "BusinessFailure" if self._is_business_failure(response) else "Success"
            )

            # Log completion
            self.logger.info(
                f"{request_type} completed in {duration_ms:.2f}ms - {status}",
                timestamp=datetime.utcnow().isoformat(),
                request_type=request_type,
                duration=duration_ms,
                status=status,
                transaction_id=self._get_transaction_id(request),
            )

            return response
        except Exception as error:
            duration_ms = (time.time() - start_time) * 1000
            status = "InfrastructureError"

            # Log error with full exception details
            self.logger.error(
                f"{request_type} failed after {duration_ms:.2f}ms",
                timestamp=datetime.utcnow().isoformat(),
                request_type=request_type,
                duration=duration_ms,
                status=status,
                exception={
                    "type": type(error).__name__,
                    "message": str(error),
                    "stack": traceback.format_exc(),
                },
                transaction_id=self._get_transaction_id(request),
            )

            # Re-raise to propagate exception
            raise

    def _get_transaction_id(self, request: TRequest) -> Optional[str]:
        """
        Get transaction_id for command executions (REQUIRED per NFR-002 and BR-002).

        Returns None for queries (no transaction).
        """
        if not self._is_command(request):
            return None

        if self.unit_of_work and self.unit_of_work.has_active_transaction:
            return str(self.unit_of_work.transaction_id)

        return None

    def _is_business_failure(self, response) -> bool:
        """Detect business failures (Result.Failure) vs successes."""
        if response is None:
            return False

        # Check if response implements Result with is_failure property
        return hasattr(response, "is_failure") and response.is_failure is True

    def _is_command(self, request) -> bool:
        """Type guard to detect command requests."""
        return isinstance(request, ICommand)


__all__ = ["TelemetryBehavior", "ILogger"]
